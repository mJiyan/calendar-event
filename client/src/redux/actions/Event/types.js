const GET_EVENTS = 'EVENT/GET_EVENTS';
const CREATE_EVENT = 'EVENT/CREATE_EVENT';
const GET_EVENTDETAILS = 'EVENT/GET_EVENTDETAILS';
const UPDATE_EVENT = 'EVENT/UPDATE_EVENT';
const REMOVE_EVENT = 'EVENT/REMOVE_EVENT';
const SET_SELECTED_EVENT = 'EVENT/SET_SELECTED_EVENT';

const actionTypes = {
  GET_EVENTS,
  CREATE_EVENT,
  GET_EVENTDETAILS,
  UPDATE_EVENT,
  REMOVE_EVENT,
  SET_SELECTED_EVENT,
};

export default actionTypes;
