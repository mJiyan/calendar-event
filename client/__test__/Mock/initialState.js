import dayjs from 'dayjs';

const initialState = {
  EventReducer: {
    events: [
      {
        _id: '619f90f7abfdf7635dc91f43',
        day: 1636531800000,
        description: 'Meet with friends',
        id: '619f90f7abfdf7635dc91f43',
        label: 'purple',
        title: 'Meeting',
      },
      {
        _id: '61cd677c3254c8001e8f0769',
        day: dayjs(),
        description: 'Buy new notebook',
        id: '61cd677c3254c8001e8f0769',
        label: 'blue',
        title: 'Notebook',
      },
    ],
    event: {},
    selectedEvent: {
      _id: "61cd8200bc6749001f5bf489",
      title: "Meeting",
      description: "Meet with friends",
      label: "purple",
      day: 1638392400000,
      __v: 0,
      id: "61cd8200bc6749001f5bf489"
    },
  },
  MonthReducer: {
    month: 1
  },
  DayReducer: {
    day: dayjs()
  },
  ModalReducer: {
    modal: false
  }
};


export default initialState;