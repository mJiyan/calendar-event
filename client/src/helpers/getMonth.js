import dayjs from 'dayjs';

export function getMonth(month = dayjs().month()) {
  const year = dayjs().year();

  const firstDayInMonth = dayjs(new Date(year, month, 0)).day();

  let presentMonth = 0 - firstDayInMonth;
  /** 5 week 7 day */
  const days2DArray = new Array(5).fill([]).map(() =>
    new Array(7).fill(null).map(() => {
      presentMonth++;
      return dayjs(new Date(year, month, presentMonth));
    }),
  );

  return days2DArray;
}
