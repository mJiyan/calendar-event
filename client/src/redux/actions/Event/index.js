import { useQuery, useMutation, useQueryClient } from 'react-query';
import { useDispatch } from 'react-redux';
import actionTypes from './types';
import apiCall from '../../../services/api';
import { HTTP } from '../../../services/constants';

export const useEventsData = () => {
  const dispatch = useDispatch();

  return useQuery(
    'events',
    async () => {
      const { data } = await apiCall('/events', null, null, HTTP.GET, null, true);
      return data;
    },
    {
      onSuccess: (data) => {
        dispatch({
          type: actionTypes.GET_EVENTS,
          payload: data,
        });
      },
      onError: (error) => {
        throw Error(error);
      },
    },
  );
};

export const useAddEventData = () => {
  const useClient = useQueryClient();
  const mutation = useMutation(
    async (event) => {
      return await apiCall('/events', event, null, HTTP.POST, null, true);
    },
    {
      onSuccess: () => {
        useClient.invalidateQueries('events');
      },
    },
  );

  return mutation;
};

export const useUpdateEventData = () => {
  const useClient = useQueryClient();
  const mutation = useMutation(
    async (event) => {
      return await apiCall(`/event/${event.id}`, event, null, HTTP.PUT, null, true);
    },
    {
      onSuccess: () => {
        useClient.invalidateQueries('events');
      },
    },
  );

  return mutation;
};

export const useDeleteEventData = () => {
  const useClient = useQueryClient();
  const mutation = useMutation(
    async (id) => {
      return await apiCall(`/event/${id}`, null, null, HTTP.DELETE, null, true);
    },
    {
      onSuccess: () => {
        useClient.invalidateQueries('events');
      },
    },
  );

  return mutation;
};

export const setSelectedEvent = (event) => async (dispatch) => {
  dispatch({
    type: actionTypes.SET_SELECTED_EVENT,
    payload: event,
  });
};
