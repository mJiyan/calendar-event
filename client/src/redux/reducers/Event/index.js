import actionTypes from '../../actions/Event/types';

const initialState = {
  event: {},
  selectedEvent: {},
  events: [],
};

const EventReducer = (state = initialState, action) => {
  switch (action.type) {
    case actionTypes.GET_EVENTS:
      return { ...state, events: action.payload };
    case actionTypes.SET_SELECTED_EVENT:
      return { ...state, selectedEvent: action.payload };
    default:
      return state;
  }
};

export default EventReducer;
