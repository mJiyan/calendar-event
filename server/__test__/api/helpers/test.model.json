{
    "event": {
        "title": "Interview",
        "description": "Interview with customer",
        "label": "red",
        "day": 1637654400000
    }
}