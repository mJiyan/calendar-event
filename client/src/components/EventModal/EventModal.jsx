import React, { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import TimePicker from '@mui/lab/TimePicker';
import TextField from '@mui/material/TextField';
import DateAdapter from '@mui/lab/AdapterDayjs';
import LocalizationProvider from '@mui/lab/LocalizationProvider';
import * as modalActions from '../../redux/actions/Modal';
import * as eventActions from '../../redux/actions/Event';
import * as dayActions from '../../redux/actions/Day';
import { useAddEventData, useDeleteEventData, useUpdateEventData } from '../../redux/actions/Event';

const labelsClasses = ['indigo', 'gray', 'green', 'blue', 'red', 'purple'];

const EventModal = () => {
  const selectedEvent = useSelector((state) => state.EventReducer.selectedEvent);
  const daySelected = useSelector((state) => state.DayReducer.day);
  const dispatch = useDispatch();

  const [title, setTitle] = useState(selectedEvent ? selectedEvent.title : '');
  const [description, setDescription] = useState(selectedEvent ? selectedEvent.description : '');
  const [selectedLabel, setSelectedLabel] = useState(
    selectedEvent ? labelsClasses.find((lbl) => lbl === selectedEvent.label) : labelsClasses[0],
  );
  const [time, setTime] = useState(
    Object.entries(selectedEvent).length !== 0 ? selectedEvent.day : daySelected,
  );

  const addMutation = useAddEventData();
  const updateMutation = useUpdateEventData();
  const removeMutation = useDeleteEventData();

  const handleSubmit = (e) => {
    e.preventDefault();
    const calendarEvent = {
      title,
      description,
      label: selectedLabel,
      day: daySelected.valueOf(),
      id: Object.entries(selectedEvent).length !== 0 ? selectedEvent.id : undefined,
    };

    try {
      if (Object.entries(selectedEvent).length !== 0) {
        updateMutation.mutate(calendarEvent);
      } else {
        addMutation.mutate(calendarEvent);
      }
      dispatch(modalActions.setShowEventModal(false));
      dispatch(eventActions.setSelectedEvent(false));
    } catch (error) {
      console.error('ERROR: ', error);
    }
  };

  return (
    <div className="h-screen w-full fixed left-0 top-0 flex justify-center items-center">
      <form className="bg-white rounded-lg shadow-2xl w-1/4">
        <header className="bg-gray-100 px-4 py-2 flex justify-between items-center">
          <span className="material-icons-outlined text-gray-400">drag_handle</span>
          <div>
            {Object.entries(selectedEvent).length !== 0 && (
              <span
                onClick={() => {
                  removeMutation.mutate(selectedEvent.id);
                  dispatch(modalActions.setShowEventModal(false));
                  dispatch(eventActions.setSelectedEvent(false));
                }}
                className="material-icons-outlined text-gray-400 cursor-pointer"
              >
                delete
              </span>
            )}
            <button
              onClick={() => {
                dispatch(modalActions.setShowEventModal(false));
                dispatch(eventActions.setSelectedEvent(false));
              }}
            >
              <span className="material-icons-outlined text-gray-400">close</span>
            </button>
          </div>
        </header>
        <div className="p-3">
          <div className="grid grid-cols-1/5 items-end gap-y-7">
            <div> </div>
            <input
              type="text"
              name="title"
              placeholder="Add title"
              value={title}
              maxLength={30}
              required
              className="pt-3 border-0 text-gray-600 text-xl font-semibold pb-2 w-full border-b-2 border-gray-200 focus:outline-none focus:ring-0 focus:border-blue-500"
              onChange={(e) => setTitle(e.target.value)}
            />
            <span className="material-icons-outlined text-gray-400">calendar_today</span>
            <p>{daySelected.format('dddd, MMMM DD')}</p>
            <span className="material-icons-outlined text-gray-400">schedule</span>
            <p>
              <LocalizationProvider dateAdapter={DateAdapter}>
                <TimePicker
                  value={time}
                  onChange={(times) => {
                    setTime(times);
                    dispatch(dayActions.setDaySelected(times));
                  }}
                  renderInput={(params) => <TextField {...params} />}
                />
              </LocalizationProvider>
            </p>
            <span className="material-icons-outlined text-gray-400">segment</span>
            <input
              type="text"
              name="description"
              placeholder="Add a description"
              value={description}
              required
              className="pt-3 border-0 text-gray-600 pb-2 w-full border-b-2 border-gray-200 focus:outline-none focus:ring-0 focus:border-blue-500"
              onChange={(e) => setDescription(e.target.value)}
            />
            <span className="material-icons-outlined text-gray-400">bookmark_border</span>
            <div className="flex gap-x-2">
              {labelsClasses.map((lblClass, i) => (
                <span
                  key={i}
                  onClick={() => setSelectedLabel(lblClass)}
                  className={`bg-${lblClass}-500 w-6 h-6 rounded-full flex items-center justify-center cursor-pointer`}
                >
                  {selectedLabel === lblClass && (
                    <span className="material-icons-outlined text-white text-sm">check</span>
                  )}
                </span>
              ))}
            </div>
          </div>
        </div>
        <footer className="flex jusify-end border-t p-3 mt-5">
          <button
            type="submit"
            onClick={handleSubmit}
            className="bg-blue-500 hover:bg-blue-600 px-6 py-2 rounded text-white"
          >
            Save
          </button>
        </footer>
      </form>
    </div>
  );
};

export default EventModal;
