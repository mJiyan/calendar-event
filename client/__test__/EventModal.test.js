import React from 'react';
import EventModal from '../src/components/EventModal/EventModal';
import { mount, shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import { getMonth } from '../src/helpers/getMonth';
import initialState from './Mock/initialState';
import { QueryClientProvider, QueryClient } from 'react-query';
import * as EventQuery from '../src/redux/actions/Event/index';
import dayjs from 'dayjs';
import toJson from 'enzyme-to-json'; //added this line


const queryClient = new QueryClient({
    defaultOptions: {
        queries: {
            retry: false,
        },
    },
})

// jest.mock("../src/redux/actions/Event/index");
// jest.spyOn(EventQuery, 'useDeleteEventData').mockResolvedValue(['mocked', 'foos']);


const mockStore = configureMockStore([thunk]);
const store = mockStore(initialState);

describe('rendering components', () => {
    const component = renderer.create(
        <QueryClientProvider client={queryClient}>
            <Provider store={store}>
                <EventModal />
            </Provider>
        </QueryClientProvider>,
    );

    const wrapperMount = mount(
        <QueryClientProvider client={queryClient}>
            <Provider store={store}>
                <EventModal />
            </Provider>
        </QueryClientProvider>,
    );

    test('it should match with the month component', () => {
        expect(component).toMatchSnapshot();
    });



    test('it should delete the event', () => {

        const deleteButton = wrapperMount.find(`[data-testid="delete-button"]`);
        deleteButton.simulate('click');

        const actions = store.getActions();
        const expectedPayload = [
            { type: 'MODAL/SET_MODAL', payload: false },
            { type: 'EVENT/SET_SELECTED_EVENT', payload: false }
        ]

        expect(actions).toEqual(expectedPayload);
    });

    test('it should close the event modal', () => {

        const closeButton = wrapperMount.find(`[data-testid="close-button"]`);
        closeButton.simulate('click');

        const actions = store.getActions();
        const expectedPayload = [
            { type: 'MODAL/SET_MODAL', payload: false },
            { type: 'EVENT/SET_SELECTED_EVENT', payload: false },
            { type: 'MODAL/SET_MODAL', payload: false },
            { type: 'EVENT/SET_SELECTED_EVENT', payload: false }
        ]


        expect(actions).toEqual(expectedPayload);
    });


    test('it should save the event', () => {

        const saveButton = wrapperMount.find(`[data-testid="save-button"]`);
        saveButton.simulate('click');

        const actions = store.getActions();
        const expectedPayload = [
            { type: 'MODAL/SET_MODAL', payload: false },
            { type: 'EVENT/SET_SELECTED_EVENT', payload: false },
            { type: 'MODAL/SET_MODAL', payload: false },
            { type: 'EVENT/SET_SELECTED_EVENT', payload: false },
            { type: 'MODAL/SET_MODAL', payload: false },
            { type: 'EVENT/SET_SELECTED_EVENT', payload: false }
        ]

        expect(actions).toEqual(expectedPayload);
    });

    test('it should save the event', () => {

        const saveButton = wrapperMount.find(`[data-testid="save-button"]`);
        saveButton.simulate('click');

        const actions = store.getActions();
        const expectedPayload = [
            { type: 'MODAL/SET_MODAL', payload: false },
            { type: 'EVENT/SET_SELECTED_EVENT', payload: false },
            { type: 'MODAL/SET_MODAL', payload: false },
            { type: 'EVENT/SET_SELECTED_EVENT', payload: false },
            { type: 'MODAL/SET_MODAL', payload: false },
            { type: 'EVENT/SET_SELECTED_EVENT', payload: false },
            { type: 'MODAL/SET_MODAL', payload: false },
            { type: 'EVENT/SET_SELECTED_EVENT', payload: false }
        ]
        expect(actions).toEqual(expectedPayload);
    });


    test('it should change the event title', () => {
        const titleInput = wrapperMount.find('[data-testid="title-input"]');
        titleInput.simulate('change', { target: { value: 'Meeting' } })
        expect(toJson(titleInput).props.value).toEqual('Meeting');
    });

    test('it should change the event description', () => {
        const descriptionInput = wrapperMount.find('[data-testid="description-input"]');
        descriptionInput.simulate('change', { target: { value: 'Meet with friends' } })
        expect(toJson(descriptionInput).props.value).toEqual('Meet with friends');
    });


    test('it should have tick after selected', () => {
        const colorButton = wrapperMount.find('[data-testid="color-button-1"]');

        colorButton.simulate('click');
        const colorButton2 = wrapperMount.find('[data-testid="color-button-1"]');
        expect(toJson(colorButton2).children.length).toEqual(1);
    });



});
