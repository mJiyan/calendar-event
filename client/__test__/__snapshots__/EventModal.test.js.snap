// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering components it should match with the month component 1`] = `
<div
  className="h-screen w-full fixed left-0 top-0 flex justify-center items-center"
  data-testid="modal-component"
>
  <form
    className="bg-white rounded-lg shadow-2xl w-1/4"
  >
    <header
      className="bg-gray-100 px-4 py-2 flex justify-between items-center"
    >
      <span
        className="material-icons-outlined text-gray-400"
      >
        drag_handle
      </span>
      <div>
        <span
          className="material-icons-outlined text-gray-400 cursor-pointer"
          data-testid="delete-button"
          onClick={[Function]}
        >
          delete
        </span>
        <button
          data-testid="close-button"
          onClick={[Function]}
        >
          <span
            className="material-icons-outlined text-gray-400"
          >
            close
          </span>
        </button>
      </div>
    </header>
    <div
      className="p-3"
    >
      <div
        className="grid grid-cols-1/5 items-end gap-y-7"
      >
        <div>
           
        </div>
        <input
          className="pt-3 border-0 text-gray-600 text-xl font-semibold pb-2 w-full border-b-2 border-gray-200 focus:outline-none focus:ring-0 focus:border-blue-500"
          data-testid="title-input"
          maxLength={30}
          name="title"
          onChange={[Function]}
          placeholder="Add title"
          required={true}
          type="text"
          value="Meeting"
        />
        <span
          className="material-icons-outlined text-gray-400"
        >
          schedule
        </span>
        <p>
          Tuesday, January 04
        </p>
        <span
          className="material-icons-outlined text-gray-400"
        >
          schedule
        </span>
        <p>
          <div
            className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          >
            <div
              className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
              onClick={[Function]}
            >
              <input
                aria-invalid={false}
                aria-label="Choose time, selected time is 12:00 AM"
                aria-readonly={true}
                autoFocus={false}
                className="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                id="mui-1"
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                readOnly={true}
                required={false}
                type="text"
                value="12:00 AM"
              />
              <fieldset
                aria-hidden={true}
                className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  className="css-nnbavb"
                >
                  <span
                    className="notranslate"
                    dangerouslySetInnerHTML={
                      Object {
                        "__html": "&#8203;",
                      }
                    }
                  />
                </legend>
              </fieldset>
            </div>
          </div>
        </p>
        <span
          className="material-icons-outlined text-gray-400"
        >
          segment
        </span>
        <input
          className="pt-3 border-0 text-gray-600 pb-2 w-full border-b-2 border-gray-200 focus:outline-none focus:ring-0 focus:border-blue-500"
          data-testid="description-input"
          name="description"
          onChange={[Function]}
          placeholder="Add a description"
          required={true}
          type="text"
          value="Meet with friends"
        />
        <span
          className="material-icons-outlined text-gray-400"
        >
          bookmark_border
        </span>
        <div
          className="flex gap-x-2"
        >
          <span
            className="bg-indigo-500 w-6 h-6 rounded-full flex items-center justify-center cursor-pointer"
            data-testid="color-button-0"
            onClick={[Function]}
          />
          <span
            className="bg-gray-500 w-6 h-6 rounded-full flex items-center justify-center cursor-pointer"
            data-testid="color-button-1"
            onClick={[Function]}
          />
          <span
            className="bg-green-500 w-6 h-6 rounded-full flex items-center justify-center cursor-pointer"
            data-testid="color-button-2"
            onClick={[Function]}
          />
          <span
            className="bg-blue-500 w-6 h-6 rounded-full flex items-center justify-center cursor-pointer"
            data-testid="color-button-3"
            onClick={[Function]}
          />
          <span
            className="bg-red-500 w-6 h-6 rounded-full flex items-center justify-center cursor-pointer"
            data-testid="color-button-4"
            onClick={[Function]}
          />
          <span
            className="bg-purple-500 w-6 h-6 rounded-full flex items-center justify-center cursor-pointer"
            data-testid="color-button-5"
            onClick={[Function]}
          >
            <span
              className="material-icons-outlined text-white text-sm"
            >
              check
            </span>
          </span>
        </div>
      </div>
    </div>
    <footer
      className="flex jusify-end border-t p-3 mt-5"
    >
      <button
        className="bg-blue-500 hover:bg-blue-600 px-6 py-2 rounded text-white"
        data-testid="save-button"
        onClick={[Function]}
        type="submit"
      >
        Save
      </button>
    </footer>
  </form>
</div>
`;
